prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,208 @@
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if len(projectQueries) == 0 {
+		return nil
+	}
+
+	// Only respond to label add or issue/PR open events
+	if !isPRChanged(pe) {
+		return nil
+	}
+
+	// Get any ProjectQueries that match this PR
+	matchedProjectQueries, err := getMatchingProjectQueries(gc, projectQueries, pe)
+	if err != nil {
+		return err
+	}
+	if len(matchedProjectQueries) == 0 {
+		return nil
+	}
+
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		le.WithError(err)
+	}
+
+	return nil
+}
+
+// labelsInLabels check that all of the queryLabels (the first list), are present in
+// the list of github.Labels (the second list)
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
+	for _, queryLabel := range queryLabels {
+		match := false
+		for _, label := range labels {
+			if queryLabel == label.Name {
+				match = true
+				break
+			}
+		}
+		if match == false {
+			return false
+		}
+	}
+	return true
+}
+
+// These are the only actions that this plugin cares about
+func isPRChanged(pe github.PullRequestEvent) bool {
+	switch pe.Action {
+	case github.PullRequestActionOpened:
+		return true
+	case github.PullRequestActionClosed:
+		return true
+	case github.PullRequestActionReopened:
+		return true
+	case github.PullRequestActionLabeled:
+		return true
+	case github.PullRequestActionUnlabeled:
+		return true
+	default:
+		return false
+	}
+}
+
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	matchedProjectQueries := make([]plugins.ProjectQuery, 0, 1)
+	// GetIssueLabels works for PRs as they are considered issues in the API
+	labels, err := gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
+	if err != nil {
+		return nil, err
+	}
+	for _, projectQuery := range projectQueries {
+		searchQuery := projectQuery.Query
+		if searchQuery.Org != "" && searchQuery.Org != pe.Repo.Owner.Login {
+			continue
+		}
+		if searchQuery.State != "" && searchQuery.State != pe.PullRequest.State {
+			continue
+		}
+		if len(searchQuery.Labels) != 0 && !labelsInLabels(searchQuery.Labels, labels) {
+			continue
+		}
+		matchedProjectQueries = append(matchedProjectQueries, projectQuery)
+	}
+	return matchedProjectQueries, nil
+}
+
+func addPRToProject(gc githubClient, projectQuery plugins.ProjectQuery, pe github.PullRequestEvent) error {
+	// Get the project id by name
+	projects, err := gc.GetOrgProjects(projectQuery.Org)
```

Comment:
For deployments without the GitHub proxy cache, the projects and columns seem like a good place to add an in-memory cache

Reply:
So would the plugin just hold onto a list of projects outside of any functions? Then it could query this first and only using `GetOrgProjects` if it can't match a project name. Or its there a particular way a cache should be implemented?

Reply:
That seems reasonable -- look at the cache in `prow/commentpruner/commentpruner.Go#EventClient.comments` for an example

Reply:
I've now added local cache of `github.Projects` and `github.ProjectColumn` to the plugin.

A struct called  `projectWithColumns` is used to hold a `github.Project` with a slice of `github.ProjectColumn`s.

The plugin holds a map of `orgRepo` strings to slices of `projectWithColumns`. Each time a PR is added to a project this map is searched first. If the project cannot be found the slice for that `orgRepo` string is updated from GitHub using `GetOrgProjects`, and the search is repeated. This process is repeated when searching for the project's column to add the PR to.

Additionally when the projects for an `orgRepo` are updated and a new slice of `projectWithColumns` is made, the columns for any project whose ID stays the same should be kept from the old slice, reducing the need to get columns from GitHub with `GetProjectColumns`. However this doesn't seem to be working currently as looking at logs `GetProjectColumns` is still called every time a PR is added to a project, so I'm still working on that.

I have also noticed I still need to add a way to get a list of GitHub projects at the repo level.